tests/integration/lib/tabs.js

```diff
@@ -2,48 +2,70 @@
 
 module.exports = function(client) {
 
-	const handles = {};
-
 	// set the current tab to the admin page.
-	const loaded = client
-	.getCurrentTabId()
-	.then(handle => handles.about = handle)
-	.newWindow('/viewer')
-	.getCurrentTabId()
-	.then(handle => handles.viewer = handle)
-	.newWindow('/admin')
-	.getCurrentTabId()
-	.then(handle => handles.admin = handle);
-
-	function loadTab(name) {
-		return loaded
-		.getCurrentTabId()
-		.then(id => {
-
-			// if you try to switch to the current tab it will switch to the
-			// first tab :/
-			if (id !== handles[name]) {
-				return loaded
-				.switchTab(handles[name]);
+	const tabs = {};
+	let loaded = client;
+
+	class Tab {
+		constructor(name, options) {
+
+			const url = options.url;
+			const handle = options.handle;
+
+			if (url) loaded = loaded
+			.newWindow(url)
+			.getCurrentTabId()
+			.then(handle => {
+				this.handle = handle;
+			});
+
+			if (handle) {
+				this.handle = handle;
 			}
-		});
-	}
-	
-	const fnInterface = {
-		admin() {
-			return loadTab('admin');
-		},
-		viewer() {
-			return loadTab('viewer');
-		},
-		about() {
-			return loadTab('about');
+
+			this.name = name;
+
+			tabs[name] = this;
+		}
+
+		ready() {
+			return loaded
+			.then(() => this);
+		}
+
+		switchTo() {
+
+			loaded = loaded
+			.getCurrentTabId()
+			.then(id => {
+				if (id !== this.handle) {
+					return client.switchTab(this.handle);
+				}
+			});
+
+			return loaded
+			.then(() => this);
 		}
+
+		close () {
+			loaded = this.switchTo()
+			.window()
+			.then(() => {
+				delete tabs[this.name];
+			})
+			.switchTab(); // go to next tab
+			return loaded;
+		}
+	}
+
+	const single = {
+		Tab,
+		tabs
 	};
 
 	module.exports = function () {
```

Comment:
This overwrites `module.exports` taking advantage of the singleton nature of commonjs modules so that it only gets instantiated once and from then on only returns the initial instantiated version.

This means we can preserve the interface across all modules which uses it and it's first use is the one which causes it to be set up.